File: 프로그래머스/lv2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.kt
import java.util.*

class Solution {
    fun solution(bridge_length: Int, weight: Int, truck_weights: IntArray): Int {
        val bridge: Queue<Int> = LinkedList(List(bridge_length) { 0 })
        val waiting: Queue<Int> = LinkedList(truck_weights.toList())
        var time = 0
        var total = 0

        while (bridge.isNotEmpty()) {
            time++
            total -= bridge.poll()

            if (waiting.isNotEmpty()) {
                if (total + waiting.peek() <= weight) {
                    total += waiting.peek()
                    bridge.add(waiting.poll())
                } else {
                    bridge.add(0)
                }
            }
        }

        return time
    }
}

File: 프로그래머스/lv2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    bridge = deque([0] * bridge_length)
    wait = deque(truck_weights)
    total_weight = 0
    time = 0
    
    while bridge:
        total_weight -= bridge.popleft()
        time += 1
        
        if wait:
            if total_weight + wait[0] <= weight:
                bridge.append(wait[0])
                total_weight += wait.popleft()
            else:
                bridge.append(0)
    
    return time